File: ios/smart_splash_kit/Sources/smart_splash_kit/SmartSplashKitPlugin.swift
import Flutter
import UIKit

/// SmartSplashKitPlugin
///
/// iOS-side plugin for Smart Splash Kit.
///
/// Responsibilities:
///  - getPlatformVersion: returns iOS version string
///  - getSystemTheme: returns "dark" or "light" based on UIUserInterfaceStyle
///  - keepSplashVisible / removeSplash: hooks for native → Flutter transition control
///
/// All splash UI, routing, animation, and logic live in Dart.
public class SmartSplashKitPlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "smart_splash_kit",
            binaryMessenger: registrar.messenger()
        )
        let instance = SmartSplashKitPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {

        case "getPlatformVersion":
            result("iOS \(UIDevice.current.systemVersion)")

        case "getSystemTheme":
            if #available(iOS 13.0, *) {
                let style = UITraitCollection.current.userInterfaceStyle
                result(style == .dark ? "dark" : "light")
            } else {
                result("light")
            }

        case "keepSplashVisible":
            // Hook: native splash is kept visible via flutter_native_splash.
            // No additional action needed on iOS.
            result(nil)

        case "removeSplash":
            // Hook: called when Flutter is ready to dismiss the native splash.
            result(nil)

        case "getDeviceInfo":
            let device = UIDevice.current
            let info: [String: Any] = [
                "os": "iOS",
                "osVersion": device.systemVersion,
                "model": device.model,
                "name": device.name,
            ]
            result(info)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/me/smart_splash_kit/SmartSplashKitPlugin.kt
package com.me.smart_splash_kit

import android.app.Activity
import android.content.Context
import android.os.Build
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * SmartSplashKitPlugin
 *
 * Android-side plugin for Smart Splash Kit.
 *
 * Responsibilities:
 *  - getPlatformVersion: returns Android OS version string
 *  - getSystemTheme: returns "dark" or "light" based on current system UI mode
 *  - keepSplashVisible / removeSplash: hooks for native → Flutter transition control
 *
 * All splash UI, routing, animation, and logic live in Dart.
 */
class SmartSplashKitPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var context: Context? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "smart_splash_kit")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${Build.VERSION.RELEASE}")
            }
            "getSystemTheme" -> {
                val currentContext = activity ?: context
                val theme = if (currentContext != null) {
                    val uiMode = currentContext.resources.configuration.uiMode and
                            android.content.res.Configuration.UI_MODE_NIGHT_MASK
                    if (uiMode == android.content.res.Configuration.UI_MODE_NIGHT_YES) "dark" else "light"
                } else {
                    "light"
                }
                result.success(theme)
            }
            "keepSplashVisible" -> {
                // Hook: called by Flutter to keep native splash visible
                // Works with flutter_native_splash – no action needed here
                // as flutter_native_splash handles this via its own mechanism.
                result.success(null)
            }
            "removeSplash" -> {
                // Hook: called by Flutter when ready to dismiss native splash
                result.success(null)
            }
            "getDeviceInfo" -> {
                val info = mapOf(
                    "os" to "Android",
                    "osVersion" to Build.VERSION.RELEASE,
                    "sdkInt" to Build.VERSION.SDK_INT,
                    "manufacturer" to Build.MANUFACTURER,
                    "model" to Build.MODEL,
                )
                result.success(info)
            }
            else -> result.notImplemented()
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {}

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
